### PR TITLE
Adding ARM64 support and optimized Neon implementation

### PR DESCRIPTION
**Pull request by RonenGvili**
This PR implements the FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_neon() and FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_neon() for ARM64 NEON.

### CODE REVIEW

**Comment by RonenGvili** [comment]
> Duplicate of #245 (which already subsumed #183 and #191) and probably incompatible with it.

Both #183 and #191 offered a partial solution regarding both the configuration and the optimization. This pr aims to offer a full solution to arm64 support with better performance.

**Comment by ktmf01** [comment]
Could you please share your performance measurements?

**Comment by RonenGvili** [comment]
> Could you please share your performance measurements?

I used Time ./flac -f -8 -s  on a wav file I have.
Original code:
       26.54 real        25.99 user         0.42 sys
Neon code:
       5.60 real         5.22 user         0.29 sys

**Review by ktmf01 — CHANGES_REQUESTED** [review]
Have you run 'make check'? It doesn't pass on my ARM64 system. It seems there is a problem somewhere in the NEON code (most likely FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_neon) that causes FLAC to behave non-lossless.

File: src/libFLAC/stream_encoder.c [hunk 1]
```diff
@@ -1063,9 +1076,17 @@ static FLAC__StreamEncoderInitStatus init_stream_internal_(
 		}
 #    endif
 #   endif /* FLAC__HAS_X86INTRIN */
-#  endif /* FLAC__CPU_... */
+#  endif /* FLAC__CPU_... */   
+
+	#if defined FLAC__CPU_ARM64
+    encoder->private_->local_lpc_compute_residual_from_qlp_coefficients_16bit = FLAC__lpc_compute_residual_from_qlp_coefficients_16_intrin_neon;
```
[the remark this answers is not in the document]
Comment (RonenGvili) [reply src/libFLAC/stream_encoder.c]: You are correct - I removed the code.
BTW - avx code has same clipping...
                                        q0  = _mm256_set1_epi32(0xffff & qlp_coeff[0 ]);
					q1  = _mm256_set1_epi32(0xffff & qlp_coeff[1 ]);
					q2  = _mm256_set1_epi32(0xffff & qlp_coeff[2 ]);
					q3  = _mm256_set1_epi32(0xffff & qlp_coeff[3 ]);
					q4  = _mm256_set1_epi32(0xffff & qlp_coeff[4 ]);
                                        ...

**Comment by RonenGvili** [comment]
> Have you run 'make check'? It doesn't pass on my ARM64 system. It seems there is a problem somewhere in the NEON code (most likely FLAC__lpc_compute_residual_from_qlp_coefficients_wide_intrin_neon) that causes FLAC to behave non-lossless.

Regarding make check - I get "All tests passed". I looked for fail and got 235 lines of "... failed as expected - OK"..

**Comment by ktmf01** [comment]
If I modify the test to not stop on the first error (replacing die with echo), I get
>FLAC__TEST_LEVEL=1 FLAC__TEST_WITH_VALGRIND=no ECHO_N="-n" ECHO_C="" ./test_replaygain.sh
Generating stream...
Testing FLAC replaygain 8000 (8000 x 1) ... replaygain.flac: -12,730003 1,394607 -12,730003 1,394607
OK
Testing FLAC replaygain 11025 (11025 x 1) ... replaygain.flac: -12,910004 1,996646 -12,910004 1,996646
OK
Testing FLAC replaygain 12000 (12000 x 1) ... replaygain.flac: -13,000000 2,000000 -13,000000 2,000000
ERROR, Expected -12.98 db instead of comment[2]: REPLAYGAIN_TRACK_GAIN=-13.00 dB
ERROR, Expected -12.98 db instead of comment[4]: REPLAYGAIN_ALBUM_GAIN=-13.00 dB
OK
Testing FLAC replaygain 16000 (16000 x 1) ... replaygain.flac: -13,269997 1,000000 -13,269997 1,000000
OK
Testing FLAC replaygain 18900 (18900 x 1) ... replaygain.flac: -13,430000 1,992239 -13,430000 1,992239
ERROR, Expected -13.41 db instead of comment[2]: REPLAYGAIN_TRACK_GAIN=-13.43 dB
ERROR, Expected -13.41 db instead of comment[4]: REPLAYGAIN_ALBUM_GAIN=-13.43 dB
OK
Testing FLAC replaygain 22050 (22050 x 1) ... replaygain.flac: -13,790001 2,057170 -13,790001 2,057170
ERROR, Expected -13.77 db instead of comment[2]: REPLAYGAIN_TRACK_GAIN=-13.79 dB
ERROR, Expected -13.77 db instead of comment[4]: REPLAYGAIN_ALBUM_GAIN=-13.79 dB
OK
Testing FLAC replaygain 24000 (24000 x 1) ... replaygain.flac: -13,849998 1,903426 -13,849998 1,903426
ERROR, Expected -13.82 db instead of comment[2]: REPLAYGAIN_TRACK_GAIN=-13.85 dB
ERROR, Expected -13.82 db instead of comment[4]: REPLAYGAIN_ALBUM_GAIN=-13.85 dB
OK
Testing FLAC replaygain 28000 (28000 x 1) ... replaygain.flac: -14,080002 1,937500 -14,080002 1,937500
ERROR, Expected -14.06 db instead of comment[2]: REPLAYGAIN_TRACK_GAIN=-14.08 dB
ERROR, Expected -14.06 db instead of comment[4]: REPLAYGAIN_ALBUM_GAIN=-14.08 dB
OK
Testing FLAC replaygain 32000 (32000 x 1) ... replaygain.flac: -14,080002 1,000000 -14,080002 1,000000
OK
Testing FLAC replaygain 36000 (36000 x 1) ... replaygain.flac: -14,129997 2,000000 -14,129997 2,000000
ERROR, Expected -14.12 db instead of comment[2]: REPLAYGAIN_TRACK_GAIN=-14.13 dB
ERROR, Expected -14.12 db instead of comment[4]: REPLAYGAIN_ALBUM_GAIN=-14.13 dB
OK
Testing FLAC replaygain 37800 (37800 x 1) ... replaygain.flac: -14,190002 2,096095 -14,190002 2,096095
ERROR, Expected -14.18 db instead of comment[2]: REPLAYGAIN_TRACK_GAIN=-14.19 dB
ERROR, Expected -14.18 db instead of comment[4]: REPLAYGAIN_ALBUM_GAIN=-14.19 dB
OK
Testing FLAC replaygain 44100 (44100 x 1) ... replaygain.flac: -14,180000 1,991328 -14,180000 1,991328
ERROR, Expected -14.17 db instead of comment[2]: REPLAYGAIN_TRACK_GAIN=-14.18 dB
ERROR, Expected -14.17 db instead of comment[4]: REPLAYGAIN_ALBUM_GAIN=-14.18 dB
OK
Testing FLAC replaygain 48000 (48000 x 1) ... replaygain.flac: -14,160004 1,605853 -14,160004 1,605853
OK
Testing FLAC replaygain 96000 (48000 x 2) ... replaygain.flac: -14,160004 1,000000 -14,160004 1,000000
OK
Testing FLAC replaygain 192000 (48000 x 4) ... replaygain.flac: -14,160004 1,000000 -14,160004 1,000000
OK

This hints on rather small differences, so no integer overflow or anything similar
> FLAC__TEST_LEVEL=1 FLAC__TEST_WITH_VALGRIND=no ECHO_N="-n" ECHO_C="" ./test_streams.sh
> [...]
> Testing 24-bit sine wave streams...
sine24-00 (--channels=1 --bps=24 -0 -l 16 --lax -m -e --sample-rate=48000): encode...ERROR during encode of sine24-00
decode...compare...cmp: einde-van-bestand in sine24-00.cmp dat leeg is
ERROR during compare of sine24-00
OK
sine24-01 (--channels=1 --bps=24 -0 -l 16 --lax -m -e --sample-rate=96000): encode...decode...compare...OK
sine24-02 (--channels=1 --bps=24 -0 -l 16 --lax -m -e): encode...ERROR during encode of sine24-02
decode...compare...cmp: einde-van-bestand in sine24-02.cmp na byte 282624, in regel 943
ERROR during compare of sine24-02
OK
sine24-03 (--channels=1 --bps=24 -0 -l 16 --lax -m -e): encode...ERROR during encode of sine24-03
decode...compare...cmp: einde-van-bestand in sine24-03.cmp na byte 12288, in regel 1
ERROR during compare of sine24-03
OK
sine24-04 (--channels=1 --bps=24 -0 -l 16 --lax -m -e): encode...ERROR during encode of sine24-04
decode...compare...cmp: einde-van-bestand in sine24-04.cmp na byte 24576, in regel 1
ERROR during compare of sine24-04
OK
sine24-10 (--channels=2 --bps=24 -0 -l 16 --lax -m -e --sample-rate=48000): encode...ERROR during encode of sine24-10
decode...compare...cmp: einde-van-bestand in sine24-10.cmp na byte 122880, in regel 381
ERROR during compare of sine24-10
OK
sine24-11 (--channels=2 --bps=24 -0 -l 16 --lax -m -e --sample-rate=48000): encode...decode...compare...OK
sine24-12 (--channels=2 --bps=24 -0 -l 16 --lax -m -e --sample-rate=96000): encode...decode...compare...OK
sine24-13 (--channels=2 --bps=24 -0 -l 16 --lax -m -e): encode...ERROR during encode of sine24-13
decode...compare...cmp: einde-van-bestand in sine24-13.cmp na byte 73728, in regel 1
ERROR during compare of sine24-13
OK
sine24-14 (--channels=2 --bps=24 -0 -l 16 --lax -m -e): encode...ERROR during encode of sine24-14
decode...compare...cmp: einde-van-bestand in sine24-14.cmp na byte 49152, in regel 1
ERROR during compare of sine24-14
OK
sine24-15 (--channels=2 --bps=24 -0 -l 16 --lax -m -e): encode...decode...compare...OK
sine24-16 (--channels=2 --bps=24 -0 -l 16 --lax -m -e): encode...ERROR during encode of sine24-16
decode...compare...cmp: einde-van-bestand in sine24-16.cmp na byte 49152, in regel 329
ERROR during compare of sine24-16
OK
sine24-17 (--channels=2 --bps=24 -0 -l 16 --lax -m -e): encode...ERROR during encode of sine24-17
decode...compare...cmp: einde-van-bestand in sine24-17.cmp na byte 565248, in regel 2544
ERROR during compare of sine24-17
OK
sine24-18 (--channels=2 --bps=24 -0 -l 16 --lax -m -e): encode...ERROR during encode of sine24-18
decode...compare...cmp: einde-van-bestand in sine24-18.cmp na byte 24576, in regel 45
ERROR during compare of sine24-18
OK
sine24-19 (--channels=2 --bps=24 -0 -l 16 --lax -m -e): encode...ERROR during encode of sine24-19
decode...compare...cmp: einde-van-bestand in sine24-19.cmp na byte 49152, in regel 83
ERROR during compare of sine24-19
OK

If I take one and try by hand, I get

>~/bin/flac-RonenGvili/test $ ../src/flac/flac -V -f --channels=2 --endian=little --sign=signed --bps=24 -0 -l 16 --lax -m -e --sample-rate=48000 sine24-10.raw -o test.flac
>
>flac 1.3.3
Copyright (C) 2000-2009  Josh Coalson, 2011-2016  Xiph.Org Foundation
flac comes with ABSOLUTELY NO WARRANTY.  This is free software, and you are
welcome to redistribute it under certain conditions.  Type `flac' for details.
>
>
>sine24-10.raw: ERROR during encoding
               state = FLAC__STREAM_ENCODER_VERIFY_MISMATCH_IN_AUDIO_DATA
>
>sine24-10.raw: ERROR: mismatch in decoded data, verify FAILED!
       Absolute sample=24575, frame=5, channel=0, sample=4095, expected -8128849, got -16124241
       In all known cases, verify errors are caused by hardware problems,
       usually overclocking or bad RAM.  Delete test.flac
       and repeat the flac command exactly as before.  If it does not give a
       verify error in the exact same place each time you try it, then there is
       a problem with your hardware; please see the FAQ:
           http://xiph.org/flac/faq.html#tools__hardware_prob
       If it does fail in the exact same place every time, keep
       tesqt.flac and submit a bug report to:
           https://sourceforge.net/p/flac/bugs/
       Make sure to upload the FLAC file and use the "Monitor" feature to
       monitor the bug status.
Verify FAILED!  Do not trust test.flac

**Comment by RonenGvili** [comment]
Using my M1 I get:
Testing FLAC replaygain 12000 (12000 x 1) ...  replaygain.flac: -12.980003 1.000000 -12.980003 1.000000
OK

Can you give me some details regarding your ARM64 platform?
I'll try get something similar device to run the make check on it.

**Comment by ktmf01** [comment]
I use a Raspberry Pi 4 with a 64-bit Raspbian image.

However, instead of acquiring a device, you can also try Travis CI. Testing on ARM is free for public repositories. If you branch your repository (as not to pollute this pull request) and [use this .travis.yml file](https://github.com/ktmf01/flac/blob/autoc-double-CI/.travis.yml) in that branch, you can run the test suite on another ARM machine.

When signing up for Travis (Travis-ci.com) make sure you disable automatic starting of testing, because the master branch of your repository (the one this pull request is based on) has a .travis.yml has x86_64 linux and Mac OS targets, and when you run these, you might deplete the free credits you get on signing up. While testing on ARM is free and will not cost you credits, they won't start if you have a negative credit balance.

**Comment by RonenGvili** [comment]
Hi ktmf01 - I managed to reproduce the issue using gcc. I am using clang and it works as expected. I am on it.

File: src/libFLAC/lpc_intrin_neon.c [hunk 2]
```diff
@@ -0,0 +1,995 @@
...
+				case 21: sum += qlp_coeff[20] * data[i-21]; /* Falls through. */
+				case 20: sum += qlp_coeff[19] * data[i-20]; /* Falls through. */
+				case 19: sum += qlp_coeff[18] * data[i-19]; /* Falls through. */
+				case 18: sum += qlp_coeff[17] * data[i-18]; /* Falls through. */
+				case 17: sum += qlp_coeff[16] * data[i-17]; /* Falls through. */
+				case 16: sum += qlp_coeff[15] * data[i-16]; /* Falls through. */
+				case 15: sum += qlp_coeff[14] * data[i-15]; /* Falls through. */
+				case 14: sum += qlp_coeff[13] * data[i-14]; /* Falls through. */
+				case 13: sum += qlp_coeff[12] * data[i-13];
+				         sum += qlp_coeff[11] * data[i-12];
```
Comment (ktmf01) [inline src/libFLAC/lpc_intrin_neon.c:971]: I think I found the issue. Compare these lines with these: https://github.com/xiph/flac/blob/b358381a102a2c1c153ee4cf95dfc04af62faa1a/src/libFLAC/lpc.c#L743-L774

Multiplication is now done in 32-bit math while it has to be 64-bit math. Same for the switch (order) above, that is the actual problem here.

**Comment by RonenGvili** [comment]
@ktmf01 I pushed a fix for the gcc issue.

**Comment by ktmf01** [comment]
I just profiled the code and it seems FLAC__lpc_compute_residual_from_qlp_coefficients and FLAC__lpc_compute_residual_from_qlp_coefficients_wide or not faster, but slower than the code without intrinsics, at least on my machine. It seems the compiler (gcc in this case) creates smarter routines than you did, sadly.

[Here you'll find a text file with profiling results.](https://github.com/xiph/flac/files/7830934/flac-RonenGvili.profile.txt) I abbreviated the tables to the slowest 8 functions. Above each table there is a short description what was tested. I compared you repository as-is and with --disable-asm-optimizations added to configure. For this comparison the third column (self seconds) is the most valuable.

As you can see, FLAC__lpc_compute_residual_from_qlp_coefficients_intrin_neon takes up more seconds in each test than FLAC__lpc_compute_residual_from_qlp_coefficients. This is most clearly visible when testing the -8ep setting, which executes these functions extensively. Here, the slowdown for the non-wide function is only 14%, for the wide function the NEON functions are twice times as slow.

Perhaps you can test this yourself by timing flac with and without your optimizations on setting -8ep, because is relies heavily on the FLAC__lpc_compute_residual_from_qlp_coefficients function.

**Comment by RonenGvili** [comment]
@ktmf01 -  The performance of the clang generated code is indeed inferior and same for the gcc autovectorized code vs my own Neon code. I am optimizing the code now to try and make it faster for both clang and gcc.

**Comment by RonenGvili** [comment]
@ktmf01 - Code was re-optimized and benchmarked again. Clang performance is now 32% better than previous commit. GCC performance is same for the (manual) vectorized code and the auto-vectorized.

**Comment by ktmf01** [comment]
I get the following speed increases of the function `FLAC__lpc_compute_residual_from_qlp_coefficients` with these intrinsics:

Compiler | Input bit-per-sample | Speed boost
:------------|:---------------------------|:-----------
GCC 10.2.1 | 16-bit |  +4%
Clang 11.0.1-2 | 16-bit |  +75%
GCC 10.2.1 | 24-bit |  +2%
Clang 11.0.1-2 | 24-bit | +9%

So, the compute_residual_from_qlp_coefficients part looks good to me. As said before, the compute_autocorrelation part conflicts with the change proposed in #245.

If you feel like even further improving this code, you could try to make good use of preprocessor directives to eliminate as much duplicate code as possible. Currently the compute_residual_from_qlp_coefficients functions with NEON intrinsics are over 2000 lines of hard to read code with few comments. This is something that in my opinion should happen with the other intrinsic routines as well. For example

>#define MULTIPLY_SAMPLE_VECTORS_WITH_QLP_COEFF(i,j,k,n,m) {           \ 
summ_0 = vmulq_laneq_s32(i, n, m); \ 
summ_1 = vmulq_laneq_s32(j, n, m); \ 
summ_2 = vmulq_laneq_s32(k, n, m); \ 
}

Then you can replace this code

> summ_0 = vmulq_laneq_s32(tmp_vec_0, qlp_coeff_1, m);
summ_1 = vmulq_laneq_s32(tmp_vec_4, qlp_coeff_1, m);
summ_2 = vmulq_laneq_s32(tmp_vec_8, qlp_coeff_1, m);

With this

> MULTIPLY_SAMPLE_VECTORS_WITH_QLP_COEFF(tmp_vec_0, tmp_vec_4, tmp_vec_8, qlp_coeff_1)

I haven't tested this code, it is just meant to explain what I mean. Perhaps you can find even better ways to do this. The main benefits are that the code becomes clearer if you choose good macro names, errors becomes easier to spot and the amount of lines of code decreases massively.

**Comment by fanc999** [comment]
Hi,

A bit wondering about this for ARM64 Windows builds...

I was seeing some GCC extensions being used in the code, so, am I correct to assume:

*  `uint32x4_t v>> int n` means `vshrq_n_s32(v, n)`
*  `uint64x2_t v>> int n` means `vshrq_n_s64(v, n)`
*  `uint32x4_t v - uint32x4_t s` means `vsubq_s32(v, s)`

Would love to get this to work on Visual Studio builds for ARM64.

Thanks!

**Comment by RonenGvili** [comment]
Hi @fanc999  - please take a look at my latest commit.

**Comment by RonenGvili** [comment]
@ktmf01  - I made the code much more compact using macros.
Regarding replacing float with double (pr #245) for FLAC__lpc_compute_autocorrelation_intrin_neon_lag - once the PR with the double will be merged I can easily update my code and replace the 3 functions.

**Comment by fanc999** [comment]
> Hi @fanc999 - please take a look at my latest commit.

Hi, @RonenGvili,

Your commits mostly worked, though I had to add macros to initialize the [u]int32x4_t via 4-member arrays, which one can't dp directly on Visual Studio, which is considered a GCCism.  I rebased my branch with the needed additions to make the code build on Visual Studio 2017 15.9.x or later for ARM64 Windows builds.

Thanks!

**Comment by NotTsunami** [comment]
The change to `autoc[]` was merged in ee18d1b892cd476ad927d169bf6d1162beefd0dd. This can be rebased now!

**Comment by ktmf01** [comment]
I will do some rebasing. Just saying, so @RonenGvili doesn't start too and one of us goes through the trouble for nothing

**Comment by ktmf01** [comment]
@RonenGvili Just one more thing before this is merged: there is no license at the top of lpc_intrin_neon.c. Could you add one like in lpc_intrin_sse2.c? As you are the creator of the file, it wouldn't be appropriate for me to add it to the file.

**Comment by RonenGvili** [comment]
Hi @ktmf01  I added the license on top of  lpc_intrin_neon.c

**Pull request merged**